### DISCUSSION TITLE
Is NullPool expected to close the Result Cursor upon closing the connection?

### DISCUSSION THREAD

**Discussion by mnchsmn** [sqlalchemy/sqlalchemy#11274] [Usage Questions] [answered]
When using the default poolclass, the following works:
```
with create_engine("postgresql://postgres:postgres@127.0.0.1:31987/test_db").begin() as conn:
    result = conn.execute("SELECT * FROM reco.user")
print(result.scalars().all())
[UUID('00000000-0000-0000-0000-000000000000'), UUID('00000000-0000-0000-0000-000000000001'), UUID('00000000-0000-0000-0000-000000000002'), UUID('2c5199fe-6170-408a-9ad2-51dabeb70b89')]
```
However when using `poolclass=NullPool`, I get `psycopg2.InterfaceError: cursor already closed`.
```
with create_engine("postgresql://postgres:postgres@127.0.0.1:31987/test_db", poolclass=NullPool).begin() as conn:
    result = conn.execute("SELECT * FROM reco.user")
print(result.scalars().all())
Traceback (most recent call last):
  File "***/.venv/lib/python3.12/site-packages/sqlalchemy/engine/cursor.py", line 991, in fetchall
    rows = dbapi_cursor.fetchall()
           ^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.InterfaceError: cursor already closed
```

Is this expected? Is it in general discouraged to iterate the result outside of the context-manager?

Thanks for your help!

**Comment by CaselIT** [comment, marked as answer]
hi,

> Is this expected? Is it in general discouraged to iterate the result outside of the context-manager?

yes and yes

**Reply by mnchsmn** [reply]
Thanks for the quick reply. Just for my understanding, why does it work with the standard pool and not the NullPool?

**Reply by CaselIT** [reply]
The queue pool does not close the connection when returning a connection to the pool and it seems that psycog2 does not invalidate the cursor when commit or rollback is called
